HeterogeneousCore/CUDACore/src/GPUCuda.cc

```diff
@@ -0,0 +1,105 @@
+    //
+    // TODO: improve. Possible ideas include
+    // - allocate M (< N(edm::Streams)) buffers per device per module, choose dynamically which (buffer, device) to use
+    //   * the first module of a chain dictates the device for the rest of the chain
+    // - our own CUDA memory allocator
+    //   * being able to cheaply allocate+deallocate scratch memory allows to make the execution fully dynamic e.g. based on current load
+    //   * would probably still need some buffer space/device to hold e.g. conditions data
+    //     - for conditions, how to handle multiple lumis per job?
+    deviceId_ = id % cudaService->numberOfDevices();
+
+    cuda::device::current::scoped_override_t<> setDeviceForThisScope(deviceId_);
+
+    // Create the CUDA stream for this module-edm::Stream pair
+    auto current_device = cuda::device::current::get();
+    cudaStream_ = std::make_unique<cuda::stream_t<>>(current_device.create_stream(cuda::stream::implicitly_synchronizes_with_default_stream));
+
+    beginStreamGPUCuda(id, *cudaStream_);
+  }
+
+  bool GPUCuda::call_acquireGPUCuda(DeviceBitSet inputLocation, edm::HeterogeneousEvent& iEvent, const edm::EventSetup& iSetup, edm::WaitingTaskWithArenaHolder waitingTaskHolder) {
+    if(!enabled_) {
+      return false;
+    }
+
+    // TODO: currently 'forced_ == true' is already assumed. When the
+    // scheduling logic evolves, add explicit treatment of forced_.
+
+    cuda::device::current::scoped_override_t<> setDeviceForThisScope(deviceId_);
+
+    try {
+      iEvent.setInputLocation(HeterogeneousDeviceId(HeterogeneousDevice::kGPUCuda, 0));
+      acquireGPUCuda(iEvent, iSetup, *cudaStream_);
+      cudaStream_->enqueue.callback([deviceId=deviceId_,
+                                     waitingTaskHolder, // copy needed for the catch block
+                                     locationSetter = iEvent.locationSetter()
+                                     ](cuda::stream::id_t streamId, cuda::status_t status) mutable {
+                                      if (status == cudaSuccess) {
+                                        locationSetter(HeterogeneousDeviceId(HeterogeneousDevice::kGPUCuda, deviceId));
+                                        LogTrace("GPUCuda") << "  GPU kernel finished (in callback) device " << deviceId << " CUDA stream " << streamId;
+                                        waitingTaskHolder.doneWaiting(nullptr);
+                                      } else {
+                                        // wrap the exception in a try-catch block to let GDB "catch throw" break on it
+                                        try {
+                                          auto error = cudaGetErrorName(status);
+                                          auto message = cudaGetErrorString(status);
+                                          throw cms::Exception("CUDAError") << "Callback of CUDA stream " << streamId << " in device " << deviceId << " error " << error << ": " << message;
```

Comment:
I think this can be replaced with `std::make_exception_ptr`.

Reply:
we were using that before,  but I changed to a `throw` in order to be able to `catch throw` it in GDB

Reply:
You could change `catch(..)` to `catch(cms::Exception const&)` instead, since you know the type.

Reply:
ah, I see what you mean - sure, that makes sense.